File: 36-valid-sudoku.py
# see js solution for thought process


class Solution:
    def isValidSudoku(self, board: List[List[str]]) -> bool:
        col1 = []
        col2 = []
        col3 = []
        col4 = []
        col5 = []
        col6 = []
        col7 = []
        col8 = []
        col9 = []

        for row in board:
            col1.append(row[0])
            col2.append(row[1])
            col3.append(row[2])
            col4.append(row[3])
            col5.append(row[4])
            col6.append(row[5])
            col7.append(row[6])
            col8.append(row[7])
            col9.append(row[8])

            stripped_row = list(filter(lambda x: x != ".", row))
            unique_row = set(stripped_row)

            if len(stripped_row) != len(unique_row):
                return False

        cols = [col1, col2, col3, col4, col5, col6, col7, col8, col9]

        for col in cols:
            stripped_col = list(filter(lambda x: x != ".", col))
            unique_col = set(stripped_col)

            if len(stripped_col) != len(unique_col):
                return False

        sub_box1 = []
        sub_box2 = []
        sub_box3 = []
        sub_box4 = []
        sub_box5 = []
        sub_box6 = []
        sub_box7 = []
        sub_box8 = []
        sub_box9 = []

        for i in range(0, 3):
            sub_box1.append(board[0][i])
            sub_box1.append(board[1][i])
            sub_box1.append(board[2][i])
            sub_box2.append(board[3][i])
            sub_box2.append(board[4][i])
            sub_box2.append(board[5][i])
            sub_box3.append(board[6][i])
            sub_box3.append(board[7][i])
            sub_box3.append(board[8][i])

        for i in range(3, 6):
            sub_box4.append(board[0][i])
            sub_box4.append(board[1][i])
            sub_box4.append(board[2][i])
            sub_box5.append(board[3][i])
            sub_box5.append(board[4][i])
            sub_box5.append(board[5][i])
            sub_box6.append(board[6][i])
            sub_box6.append(board[7][i])
            sub_box6.append(board[8][i])

        for i in range(6, 9):
            sub_box7.append(board[0][i])
            sub_box7.append(board[1][i])
            sub_box7.append(board[2][i])
            sub_box8.append(board[3][i])
            sub_box8.append(board[4][i])
            sub_box8.append(board[5][i])
            sub_box9.append(board[6][i])
            sub_box9.append(board[7][i])
            sub_box9.append(board[8][i])

        sub_boxes = [sub_box1, sub_box2, sub_box3, sub_box4, sub_box5, sub_box6, sub_box7, sub_box8, sub_box9]

        for sub_box in sub_boxes:
            stripped_sub_box = list(filter(lambda x: x != ".", sub_box))
            unique_sub_box = set(stripped_sub_box)

            if len(stripped_sub_box) != len(unique_sub_box):
                return False

        return True

File: 36-valid-sudoku.js
/*
Determine if a 9 x 9 Sudoku board is valid. Only the filled cells need to be validated according to the following rules:

Each row must contain the digits 1-9 without repetition.
Each column must contain the digits 1-9 without repetition.
Each of the nine 3 x 3 sub-boxes of the grid must contain the digits 1-9 without repetition.
Note:

A Sudoku board (partially filled) could be valid but is not necessarily solvable.
Only the filled cells need to be validated according to the mentioned rules.
*/

/*
Example 1:
Input: board =
[["5","3",".",".","7",".",".",".","."]
,["6",".",".","1","9","5",".",".","."]
,[".","9","8",".",".",".",".","6","."]
,["8",".",".",".","6",".",".",".","3"]
,["4",".",".","8",".","3",".",".","1"]
,["7",".",".",".","2",".",".",".","6"]
,[".","6",".",".",".",".","2","8","."]
,[".",".",".","4","1","9",".",".","5"]
,[".",".",".",".","8",".",".","7","9"]]
Output: true

Example 2:
Input: board =
[["8","3",".",".","7",".",".",".","."]
,["6",".",".","1","9","5",".",".","."]
,[".","9","8",".",".",".",".","6","."]
,["8",".",".",".","6",".",".",".","3"]
,["4",".",".","8",".","3",".",".","1"]
,["7",".",".",".","2",".",".",".","6"]
,[".","6",".",".",".",".","2","8","."]
,[".",".",".","4","1","9",".",".","5"]
,[".",".",".",".","8",".",".","7","9"]]
Output: false
Explanation: Same as Example 1, except with the 5 in the top left corner being modified to 8. Since there are two 8's in the top left 3x3 sub-box, it is invalid.
*/

/*
1.) check rows: strip row of .s, put into a set, compare stripped length to size of set
2.) check columns: same method as rows, put each column value into a new array per column
3.) check 3x3s: same method as rows, put each 3x3 section into a new array
4.) if all checks pass return true
*/

// SLOW: O(n^2)
// space inefficient: O(n) (27 arrays of length 9)

var isValidSudoku = function(board) {
    let col1 = [];
    let col2 = [];
    let col3 = [];
    let col4 = [];
    let col5 = [];
    let col6 = [];
    let col7 = [];
    let col8 = [];
    let col9 = [];

    for (let row of board) {
        col1.push(row[0]);
        col2.push(row[1]);
        col3.push(row[2]);
        col4.push(row[3]);
        col5.push(row[4]);
        col6.push(row[5]);
        col7.push(row[6]);
        col8.push(row[7]);
        col9.push(row[8]);

        let strippedRow = row.filter((num) => num !== ".");
        let uniqueRow = new Set(strippedRow);
        if (strippedRow.length !== uniqueRow.size) return false;
    }

    let cols = [col1, col2, col3, col4, col5, col6, col7, col8, col9];

    for (let col of cols) {
        console.log(`col`, col);
        let strippedCol = col.filter((num) => num !== ".");
        let uniqueCol = new Set(strippedCol);
        if (strippedCol.length !== uniqueCol.size) return false;
    }

    let subBox1 = [];
    let subBox2 = [];
    let subBox3 = [];
    let subBox4 = [];
    let subBox5 = [];
    let subBox6 = [];
    let subBox7 = [];
    let subBox8 = [];
    let subBox9 = [];

    for (let i = 0; i < 3; i++) {
        subBox1.push(board[0][i], board[1][i], board[2][i]);
        subBox2.push(board[3][i], board[4][i], board[5][i]);
        subBox3.push(board[6][i], board[7][i], board[8][i]);
    }

    for (let i = 3; i < 6; i++) {
        subBox4.push(board[0][i], board[1][i], board[2][i]);
        subBox5.push(board[3][i], board[4][i], board[5][i]);
        subBox6.push(board[6][i], board[7][i], board[8][i]);
    }

    for (let i = 6; i < 9; i++) {
        subBox7.push(board[0][i], board[1][i], board[2][i]);
        subBox8.push(board[3][i], board[4][i], board[5][i]);
        subBox9.push(board[6][i], board[7][i], board[8][i]);
    }

    let subBoxes = [subBox1, subBox2, subBox3, subBox4, subBox5, subBox6, subBox7, subBox8, subBox9];

    for (let subBox of subBoxes) {
        console.log(`subBox`, subBox);
        let strippedSubBox = subBox.filter((num) => num !== ".");
        let uniqueSubBox = new Set(strippedSubBox);
        if (strippedSubBox.length !== uniqueSubBox.size) return false;
    }

    return true;

};
